staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+				"set.all(e, e < 10)",
+				"size(set.filter(e, e%2 == 0)) == 2",
+				"set.map(e, e * 20).filter(e, e > 50).exists_one(e, e == 60)",
+				"size(set) == 5",
+			},
+		},
+		{name: "typemeta and objecmeta access specified",
+			obj: map[string]interface{}{
+				"apiVersion": "v1",
+				"kind":       "Pod",
+				"metadata": map[string]interface{}{
+					"name":         "foo",
+					"generateName": "pickItForMe",
+					"namespace":    "xyz",
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"kind":       stringType,
+				"apiVersion": stringType,
+				"metadata": *objectTypePtr(map[string]schema.Structural{
+					"name":         stringType,
+					"generateName": stringType,
+				}),
+			}),
+			valid: []string{
+				"kind == 'Pod'",
+				"apiVersion == 'v1'",
+				"metadata.name == 'foo'",
+				"metadata.generateName == 'pickItForMe'",
+			},
+			errors: map[string]string{
+				"has(metadata.namespace)": "undefined field 'namespace'",
+			},
+		},
+		{name: "typemeta and objecmeta access not specified",
+			obj: map[string]interface{}{
+				"apiVersion": "v1",
+				"kind":       "Pod",
+				"metadata": map[string]interface{}{
+					"name":         "foo",
+					"generateName": "pickItForMe",
+					"namespace":    "xyz",
+				},
+				"spec": map[string]interface{}{
+					"field1": "a",
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"spec": *objectTypePtr(map[string]schema.Structural{
+					"field1": stringType,
+				}),
+			}),
+			valid: []string{
+				"kind == 'Pod'",
+				"apiVersion == 'v1'",
+				"metadata.name == 'foo'",
+				"metadata.generateName == 'pickItForMe'",
+				"spec.field1 == 'a'",
+			},
+			errors: map[string]string{
+				"has(metadata.namespace)": "undefined field 'namespace'",
+			},
+		},
+
+		// Kubernetes special types
+		{name: "embedded object",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": *objectTypePtr(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.kind == 'Pod'",
+				"embedded.apiVersion == 'v1'",
+				"embedded.metadata.name == 'foo'",
+				"embedded.metadata.generateName == 'pickItForMe'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
```

Comment:
this field isn't tested, we could either remove it or add a new error scenario similar to the one above.